automap/src/comm_layer/pmp.rs

```diff
@@ -396,6 +448,139 @@ impl ThreadGuts {
         }
     }
 
+    fn parse_buffer(
+        &self,
+        buffer: &[u8],
+        source_address: SocketAddr,
+    ) -> Result<Ipv4Addr, AutomapError> {
+        match PmpPacket::try_from(buffer) {
+            Ok(packet) => {
+                if packet.direction != Direction::Response {
+                    let err_msg = format!(
+                        "Unexpected PMP Get request (request!) from router at {}: ignoring",
+                        source_address
+                    );
+                    warning!(&self.logger, "{}", err_msg);
+                    return Err(AutomapError::ProtocolError(err_msg));
+                }
+                if packet.opcode == Opcode::Get {
+                    let opcode_data = packet
+                        .opcode_data
+                        .as_any()
+                        .downcast_ref::<GetOpcodeData>()
+                        .expect("A Get opcode shouldn't parse anything but GetOpcodeData");
+                    Ok(opcode_data
+                        .external_ip_address_opt
+                        .expect("A Response should always produce an external ip address"))
+                } else {
+                    let err_msg = format!(
+                        "Unexpected PMP {:?} response (instead of Get) from router at {}: ignoring",
+                        packet.opcode, source_address
+                    );
+                    warning!(&self.logger, "{}", err_msg);
+                    Err(AutomapError::ProtocolError(err_msg))
+                }
+            }
+            Err(e) => {
+                error!(
+                    &self.logger,
+                    "Unparseable PMP packet ({:?}):\n{}",
+                    &e,
+                    PrettyHex::hex_dump(&buffer)
+                );
+                let err_msg = format!(
+                    "Unparseable packet ({:?}) from router at {}: ignoring",
+                    &e, source_address
+                );
+                warning!(
+                    &self.logger,
+                    "{}\n{}",
+                    err_msg,
+                    PrettyHex::hex_dump(&buffer)
+                );
+                Err(AutomapError::ProtocolError(err_msg))
+            }
+        }
+    }
+
+    fn handle_announcement(&self, public_ip: Ipv4Addr, mapping_config_opt: &Option<MappingConfig>) {
```

Comment:
Theoretically, you could have prepared this for IpAddr, allowing both Ipv4 and Ipv6. Of course I know Node cannot handle Ipv6 and is not gonna do it in a foreseeable future but at some time we'll probably need to implement it. Never mind... 

...I think I understand why you strive to keep away from the more general data structure, as you Americans say, it is pain in the butt.

Reply:
The PMP protocol is IPv4-only. I can't use IPv6 even if I want.